clang/lib/CodeGen/CodeGenModule.cpp

```diff
@@ -4303,21 +4288,15 @@ void CodeGenModule::emitMultiVersionFunctions() {
                 continue;
 
               llvm::Function *Func = createFunction(CurFD, I);
-              StringRef Architecture;
               Feats.clear();
-              if (getTarget().getTriple().isAArch64())
-                TC->getFeatures(Feats, I);
-              else if (getTarget().getTriple().isRISCV()) {
-                StringRef Version = TC->getFeatureStr(I);
-                Feats.push_back(Version);
+              if (getTarget().getTriple().isX86()) {
+                TC->getAddedFeatures(Feats, I);
```

Comment:
Yes, as I said earlier the struct MultiVersionResolverOption (see https://github.com/llvm/llvm-project/pull/116257/files#diff-731858fb71c12faa05d9549cf109d33744fa2e070c06d1740b9020c66e7d143eL5335) was originally tailored to x86. The `Architecture` field is used here: https://github.com/llvm/llvm-project/pull/116257/files#diff-9f23818ed51d0b117b5692129d0801721283d0f128a01cbc562353da0266d7adL2842.

Reply:
Do you mean in the first case, `I` actually points to `arch=sandybridge,default,sse4.2` rather than `arch=sandybridge`? If it is true, I think that's something we need to pre-process in general instead of specialize to X86. X86 just needs to handle `Architecture` here.

Reply:
No, `I` goes from 0 to 2 in your example. The proplem with these parsing functions in the .td file is that they are shared among all targets but they are not virtual. Each target has slightly different semantics though. If you look at line https://github.com/llvm/llvm-project/pull/116257/files#diff-e724febedab9c1a2832bf2056d208ff02ddcb2e6f90b5a653afc9b19ac78a5d7L4283 I am basically adding the same functions to the `target_clones` attribute that are used for the `target` attribute, so that the code is cleaner across targets, hiding target-specific parsing details from CodeGenModule::emitMultiVersionFunctions.

Reply:
Go it. Fixed in latest revision.